python/ray/experimental/channel/torch_tensor_nccl_channel.py

```diff
@@ -25,6 +26,135 @@
 logger = logging.getLogger(__name__)
 
 
+class NestedTorchTensorNcclChannel(ChannelInterface):
+    def __init__(
+        self,
+        writer: ray.actor.ActorHandle,
+        readers: List[ray.actor.ActorHandle],
+        gpu_data_typ: "TorchTensorType",
+        cpu_data_typ: Optional["SharedMemoryType"] = None,
+        _gpu_data_channel: Optional["TorchTensorNcclChannel"] = None,
+        _cpu_data_channel: Optional["Channel"] = None,
+    ):
+        self._writer = writer
+        self._readers = readers
+
+        if _gpu_data_channel is not None or _cpu_data_channel is not None:
+            # This path is used when the NestedTorchTensorNcclChannel is being
+            # deserialized.
+            assert (
+                writer is None
+                and readers is None
+                and gpu_data_typ is None
+                and cpu_data_typ is None
+            )
+            assert _gpu_data_channel is not None and _cpu_data_channel is not None
+            self._gpu_data_channel = _gpu_data_channel
+            self._cpu_data_channel = _cpu_data_channel
+        else:
+            # This path is used when the NestedTorchTensorNcclChannel is first
+            # being created, by the writer of the channel.
+            self._gpu_data_channel: TorchTensorNcclChannel = (
+                gpu_data_typ.create_channel(writer, readers)
+            )
+            self._cpu_data_channel: Optional["Channel"] = None
```

Comment:
why don't we always just create cpu channel for gpu (and remove this nested nccl channel)? Does it affect the performance drastically?

Reply:
In my understanding, the GPU channel (i.e., `TorchTensorNcclChannel`) already has two channels:

* NCCL channel: used to transfer tensors.
* CPU channel: used to transfer the tensor's metadata (i.e., shape).

The additional CPU channel here is used to transfer CPU data, which is not the tensor's metadata. Hence, we have two CPU channels, one for tensor's metadata and one for other CPU data.

> Does it affect the performance drastically?

To the best of my knowledge, the answer is no.

Reply:
Right now the performance hit for dynamic shape compared to NCCL is about 2x. On top of that, the additional CPU channel for nested tensors is about 1.5x. This is a bit higher than expected, though (see #45319).

Reply:
> Right now the performance hit for dynamic shape compared to NCCL is about 2x.

Do you mean

* Case 1: Send tensors via NCCL channel (only tensors, no tensor's metadata)
* Case 2:
  * Send tensors via NCCL channel
  * Send metadata via `meta_channel` (CPU channel)
  * Send other CPU data via `cpu_data_channel` (CPU channel)

(Case 1 performance) / (Case 2 performance) = 2?


> On top of that, the additional CPU channel for nested tensors is about 1.5x.

Do you mean

* Case 1:
  * Send tensors via NCCL channel
  * Send metadata and other CPU data via `meta_channel` (CPU channel)
* Case 2:
  * Send tensors via NCCL channel
  * Send metadata via `meta_channel` (CPU channel)
  * Send other CPU data via `cpu_data_channel` (CPU channel)

(Case 1 performance) / (Case 2 performance) = 1.5?

Thanks!

Reply:
For the first one, the comparison is between these:
```
Case 1: Send tensors via NCCL channel (only tensors, no tensor's metadata)
Case 2:

    Send tensors via NCCL channel
    Send metadata via meta_channel (CPU channel)
```

The second one is right.

Reply:
> Case 1: Send tensors via NCCL channel (only tensors, no tensor's metadata)
> Case 2:
>   *  Send tensors via NCCL channel
>   *  Send metadata via meta_channel (CPU channel)

In this case, (Case 1 performance) / (Case 2 performance) = 2. I am wondering that does it mean that:

(1)
2 * Time(Send tensors via NCCL channel) = Time(Send tensors via NCCL channel) + Time(Send metadata via meta_channel)

or 

(2)
2 * Time(Send tensors via NCCL channel) = Max(Time(Send tensors via NCCL channel), Time(Send metadata via meta_channel))

It seems to be (1) because the call is not async. Is my understanding correct?